contracts/DomainRegistry.sol

```diff
@@ -11,55 +16,99 @@ contract DomainRegistry {
 
     mapping(string => Domain) public domains;
 
-    event DomainCreated(string indexed topLevelDomain, uint256 deposit, address indexed owner);
-    event DomainReleased(string indexed topLevelDomain);
-
+    event DomainCreated(string indexed domainName, uint256 deposit, address indexed owner);
+    event DomainReleased(string indexed domainName);
 
     modifier hasRequiredDeposit() {
         require(msg.value == REQUIRED_DEPOSIT, "Wrong eth amount");
         _;
     }
 
-    modifier domainDoesNotExist(string memory _topLevelDomain) {
-        require(!domains[_topLevelDomain].isRegistered, "Domain exists");
+ modifier domainExistence(string memory domainName, bool shouldExist) {
+    bool exists = domains[domainName.stripProtocol()].owner != address(0);
+    if (shouldExist) {
+        require(exists, "Domain doesn't exist");
+    } else {
+        require(!exists, "Domain exists");
+    }
+    _;
+}
+
+    modifier isValidDomain(string memory _domainName) {
+        _domainName = _domainName.stripProtocol();
+        require(bytes(_domainName).length > 0, "Domain is empty");
+        bytes memory domainBytes = bytes(_domainName);
+        require(domainBytes[domainBytes.length - 1] != bytes1('.'), "Domain ends with a dot");
         _;
     }
 
-    modifier domainExists(string memory _topLevelDomain) {
-        require(domains[_topLevelDomain].isRegistered, "Domain doesn't exist");
+    modifier domainOwnedBySender(string memory domainName) {
+        domainName = domainName.stripProtocol();
+        require(domains[domainName].owner == msg.sender, "Not the domain owner");
         _;
     }
 
-    modifier domainOwnedBySender(string memory _topLevelDomain) {
-        require(domains[_topLevelDomain].owner == msg.sender, "Not the domain owner");
+modifier parentDomainExists(string memory domainName) {
+    if (isTopLevelDomain(domainName)) {
         _;
+        return;
+    }
+    
+    string memory parentDomain = domainName.extractParentDomain();
+    if (bytes(parentDomain).length > 1) {  
+        require(domains[parentDomain].isRegistered, "Parent domain doesn't exist");
```

Comment:
You still have this field but could've just check if `owner != address(0)` -1 point

Reply:
fxd